File: sdks/python/src/arsox_sdk/turn.py
# Copyright © 2026 Jalapeno Labs

"""A handle to one turn."""

from __future__ import annotations

import asyncio

from arsox_sdk.constants import RESULT_POLL_SECONDS
from arsox_sdk.errors import ArsoxError
from arsox_sdk.proto.arsox.turn.v1 import result_pb2, turn_pb2
from arsox_sdk.transport import Connection


class TurnHandle:
    """A handle to one turn."""

    def __init__(self, connection: Connection, thread_id: str, turn: turn_pb2.Turn) -> None:
        """Wrap a queued turn. Obtained from `ThreadHandle.start_turn`."""
        self._connection = connection
        self._thread_id = thread_id
        self.id = turn.turn_id

        # The turn as it was when queued.
        self.queued = turn

    async def get(self) -> turn_pb2.Turn:
        """Read the turn's current state.

        Raises:
            ArsoxError: when the turn is unknown or the satellite is unreachable.
        """
        response = await self._connection.call(
            "GET",
            f"/v1/threads/{self._thread_id}/turns/{self.id}",
            result_pb2.GetTurnResponse,
        )

        if not response.HasField("turn"):
            raise ArsoxError.transport("the satellite returned a turn with no turn in it")

        return response.turn

    async def result(self) -> result_pb2.TurnResult:
        """Wait for the turn to reach a terminal state and return its result.

        Polls rather than watching the stream, because a caller awaiting a result
        has not necessarily subscribed and should not have to.

        Raises:
            ArsoxError: when the turn is unknown, the satellite is unreachable,
                or a finished turn carries no result.
        """
        while True:
            response = await self._connection.call(
                "GET",
                f"/v1/threads/{self._thread_id}/turns/{self.id}",
                result_pb2.GetTurnResponse,
            )

            status = response.turn.status if response.HasField("turn") else turn_pb2.TURN_STATUS_UNSPECIFIED
            if status not in (turn_pb2.TURN_STATUS_QUEUED, turn_pb2.TURN_STATUS_RUNNING):
                if not response.HasField("result"):
                    raise ArsoxError.transport(
                        "the satellite finished a turn without recording a result"
                    )

                return response.result

            await asyncio.sleep(RESULT_POLL_SECONDS)

    async def cancel(self) -> turn_pb2.Turn:
        """Ask the satellite to stop this turn.

        A running turn is asked to stop cooperatively first. Work already
        committed to a branch survives either way.

        Raises:
            ArsoxError: when the turn is unknown or the satellite is unreachable.
        """
        response = await self._connection.call(
            "POST",
            f"/v1/threads/{self._thread_id}/turns/{self.id}/cancel",
            turn_pb2.CancelTurnResponse,
        )

        if not response.HasField("turn"):
            raise ArsoxError.transport("the satellite cancelled a turn without saying so")

        return response.turn

File: sdks/node/src/turn.ts
// Copyright © 2026 Jalapeno Labs

import type { Turn } from './proto/arsox/turn/v1/turn_pb.js'
import type { TurnResult } from './proto/arsox/turn/v1/result_pb.js'
import type { Connection } from './transport.js'

// Core
import { setTimeout as sleep } from 'node:timers/promises'

// Misc
import { CancelTurnResponseSchema } from './proto/arsox/turn/v1/turn_pb.js'
import { GetTurnResponseSchema } from './proto/arsox/turn/v1/result_pb.js'
import { TurnStatus } from './proto/arsox/turn/v1/turn_pb.js'
import { ArsoxError } from './error.js'
import { RESULT_POLL_MILLISECONDS } from './constants.js'

/** A handle to one turn. */
export class TurnHandle {
  readonly id: string

  /** The turn as it was when queued. */
  readonly queued: Turn

  readonly #connection: Connection
  readonly #threadId: string

  constructor(connection: Connection, threadId: string, turn: Turn) {
    this.#connection = connection
    this.#threadId = threadId
    this.id = turn.turnId
    this.queued = turn
  }

  /** Reads the turn's current state. */
  async get(): Promise<Turn> {
    const response = await this.#connection.call(
      'GET',
      `/v1/threads/${this.#threadId}/turns/${this.id}`,
      GetTurnResponseSchema
    )

    if (!response.turn) {
      throw ArsoxError.transport('the satellite returned a turn with no turn in it')
    }

    return response.turn
  }

  /**
   * Waits for the turn to reach a terminal state and returns its result.
   *
   * Polls rather than watching the stream, because a caller awaiting a result
   * has not necessarily subscribed and should not have to.
   */
  async result(): Promise<TurnResult> {
    for (;;) {
      const response = await this.#connection.call(
        'GET',
        `/v1/threads/${this.#threadId}/turns/${this.id}`,
        GetTurnResponseSchema
      )

      const status = response.turn?.status ?? TurnStatus.UNSPECIFIED
      if (status !== TurnStatus.QUEUED && status !== TurnStatus.RUNNING) {
        if (!response.result) {
          throw ArsoxError.transport(
            'the satellite finished a turn without recording a result'
          )
        }
        return response.result
      }

      await sleep(RESULT_POLL_MILLISECONDS)
    }
  }

  /**
   * Asks the satellite to stop this turn.
   *
   * A running turn is asked to stop cooperatively first. Work already committed
   * to a branch survives either way.
   */
  async cancel(): Promise<Turn> {
    const response = await this.#connection.call(
      'POST',
      `/v1/threads/${this.#threadId}/turns/${this.id}/cancel`,
      CancelTurnResponseSchema
    )

    if (!response.turn) {
      throw ArsoxError.transport('the satellite cancelled a turn without saying so')
    }

    return response.turn
  }
}
